src/OrbitCaptureClient/ApiEventProcessor.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitCaptureClient/ApiEventProcessor.h"
+
+#include "OrbitBase/Logging.h"
+
+using orbit_client_protos::TimerInfo;
+using orbit_grpc_protos::ApiEvent;
+
+ApiEventProcessor::ApiEventProcessor(CaptureListener* listener) : capture_listener_(listener) {}
+
+static inline TimerInfo TimerInfoFromEncodedEvent(const orbit_api::EncodedEvent& encoded_event,
+                                                  uint64_t start, uint64_t end, int32_t pid,
+                                                  int32_t tid, uint32_t depth) {
+  TimerInfo timer_info;
+  timer_info.set_start(start);
+  timer_info.set_end(end);
+  timer_info.set_process_id(pid);
+  timer_info.set_thread_id(tid);
+  timer_info.set_depth(depth);
+  timer_info.set_type(TimerInfo::kApiEvent);
+  timer_info.mutable_registers()->Reserve(6);
+  timer_info.add_registers(encoded_event.args[0]);
+  timer_info.add_registers(encoded_event.args[1]);
+  timer_info.add_registers(encoded_event.args[2]);
+  timer_info.add_registers(encoded_event.args[3]);
+  timer_info.add_registers(encoded_event.args[4]);
+  timer_info.add_registers(encoded_event.args[5]);
+  return timer_info;
+}
+
+void ApiEventProcessor::ProcessApiEvent(const orbit_grpc_protos::ApiEvent& grpc_api_event) {
+  orbit_api::ApiEvent api_event;
+  api_event.pid = grpc_api_event.pid();
+  api_event.tid = grpc_api_event.tid();
+  api_event.timestamp_ns = grpc_api_event.timestamp_ns();
+  api_event.encoded_event.args[0] = grpc_api_event.r0();
+  api_event.encoded_event.args[1] = grpc_api_event.r1();
+  api_event.encoded_event.args[2] = grpc_api_event.r2();
+  api_event.encoded_event.args[3] = grpc_api_event.r3();
+  api_event.encoded_event.args[4] = grpc_api_event.r4();
+  api_event.encoded_event.args[5] = grpc_api_event.r5();
+  ProcessApiEvent(api_event);
+}
+
+void ApiEventProcessor::ProcessApiEvent(const orbit_api::ApiEvent& api_event) {
+  orbit_api::EventType event_type = api_event.Type();
+
+  switch (event_type) {
+    case orbit_api::kScopeStart:
+      ProcessStartEvent(api_event);
+      break;
+    case orbit_api::kScopeStop:
+      ProcessStopEvent(api_event);
+      break;
+    case orbit_api::kScopeStartAsync:
+      ProcessAsyncStartEvent(api_event);
+      break;
+    case orbit_api::kScopeStopAsync:
+      ProcessAsyncStopEvent(api_event);
+      break;
+    case orbit_api::kTrackInt:
+    case orbit_api::kTrackInt64:
+    case orbit_api::kTrackUint:
+    case orbit_api::kTrackUint64:
+    case orbit_api::kTrackFloat:
+    case orbit_api::kTrackDouble:
+    case orbit_api::kString:
+      ProcessTrackingEvent(api_event);
+      break;
+    case orbit_api::kNone:
+      ERROR("Received event of type orbit_api::kNone");
+      break;
+    default:
```

Comment:
Yep, I had made a mental note to fix this, but didn't, thank you and Florian for pointing this out, fixed now with UNREACHABLE.

Reply:
UNREACHABLE is generating runtime abort if someone adds a new value, whereas omitting default results in compile-time error. The difference is that former could be missed because it needs for new event happen in a test. Latter is much harder to miss.

Reply:
yes, good point, changed it to omit default, thanks for the precision.